### ISSUE TITLE
Mapping ou annotations doctrine 

### ISSUE THREAD

**Issue by jbouzekri** [blanc-frederic/demo-hexa#2]
J'ai regardé la vidéo youtube de ta conférence. Bravo et merci. Mais je m'interroge sur l'approche que tu aurais retenu pour un adapter doctrine afin de stocker tes decks et cartes.

1. Est ce que les annotations doctrine font parties des exceptions autorisées dans le domaine d'après toi ?
2. Quid de la gestion des relations avec le type forcé en array dans ton entité deck?
3. Quid de deckcomponent qui deviendrait une entité doctrine et vivrait dans le namespace entité du coup ? Ou multiples namespaces d'entités doctrine ?

Note: je n'imagine pas faire une transformation d'une entité doctrine vers une entité de domaine avec une interface quasi identique dans l'adapter doctrine. En termes de performances ça serait loin d'être idéal 

Merci d'avance pour ton opinion

**Comment by blanc-frederic** [comment]
Bonjour et merci pour ce commentaire

Je suis en train de préparer une implementation avec doctrine, car cela suscite des questions récurrentes. Je viendrais compléter ici dès qu'elle sera merged

1. Non, par définition elles font appel au namespace Doctrine et en sont dépendantes. C'est exactement ce que nous voulons fuir en isolant le code métier dans cette architecture. Donc surtout pas d'exception sur les annotations. Il y a dans tous les ORM de quoi paramétrer le mapping à un autre endroit (dans le dossier config/ tout naturellement)

Je ne fais pas ici le procés des annotations. Si on veut utiliser des annotations Symfony dans les controlleurs, qui sont donc dans Infrastructure\Symfony\Controller alors c'est tout à fait autorisé, puisque la dépendance à l'infra, et notamment à Symfony est déjà exprimée dans l'architecture

2. J'ai pris le type array pour "faire simple", mais effectivement une classe de Collection pourrait améliorer le code et permettre de se rapprocher d'un fonctionnement d'une entité d'un ORM. Cependant je n'utiliserai pas celle de Doctrine, car il y aurait là une dépendance dans le code métier, ce que l'on se refuse.
Ceci dit, on peut aussi passer à iterable, ou créer une "entité technique" dans Infrastructure, qui pourrait alors extends l'entité du métier et se compléter de détails techniques comme la collection de doctrine

3. Deckcomponent pourrait (devrait même, je ne vois plus trop pourquoi je ne l'avais pas fait) dans tous les cas devenir une entité métier. Dans ce cas on pourrait alors l'utiliser dans le mapping (dans config)

Il est tout à fait envisageable d'avoir plusieurs namespace d'entités, pour séparer entités techniques et entités métier, ou même pour séparer plusieurs contextes si l'application devenait trop grande

Note : on peut tout à fait envisager des transferts de données d'une entité infra à une entité domaine, cela fait partie de l'architecture. Comme toujours, tout dépend de l'utilisation qui en est faite

**Comment by jbouzekri** [comment]
Merci beaucoup pour ta réponse.

Je pense que la classe Collection/ArrayCollection est "autorisable" vu que c'est une simple classe utilitaire sans autre dépendance avec le reste de la librairie Doctrine mais c'est vrai que philosophiquement c'est à éviter.

Je reste néanmoins très partagé sur cette approche hexagonale dans SF et PHP et surtout sur le sujet des entités Doctrine et des relations. Cette implémentation exemple que tu proposes avec Doctrine sera la bienvenue pour mieux se projeter et comprendre comment tu l'as implémenté dans tes projets. Pourras tu intégrer un exemple avec une relation (manyToOne, manyToMany, ...) ? 

1. Je ne remets pas du tout en cause les bienfaits de l'architecture hexa mais le trade-off en termes de performance peut parfois empêcher cette approche de transformer l'entité infra en entité domaine. L'hydratation objet de Doctrine est déjà très chère d'un point de vue ressources. Sur une requête de liste, on la paye une première fois. Si on ajoute une boucle pour hydrater un objet domaine, une deuxième fois (encore plus si on a des sous objets). Puis souvent une troisième boucle pour normaliser/sérialiser une réponse en JSON (dans un contexte API). Ca commence à faire beaucoup de boucles et beaucoup de transformations "juste" pour respecter notre découpage et la non-dépendance entre les couches.

2. Doctrine abstrait plusieurs opérations sur les relations avec ses options cascade persist, orphan removals, ... Si on passe d'une entité infra (read), à une entité domaine (modify) pour revenir à celle infra (write), dans la dernière étape, il faut remapper notre entité et ses sous objets avec celles connues de l'unit-of-work Doctrine. Lien qu'il aura "perdu" entre temps. Il me semble que ça complexifie grandement le code (ce qui est normal quand on découpe mais ça se paye aussi en termes de risque de bugs et autre)

Donc globalement, j'ai l'impression que si on fait cette architecture, Doctrine (et un ORM en général) ne semble plus trop adapté mais donc on va se retrouver à quasiment recoder un ORM maison (au moins pour faire l'hydratation des objets domaines à partir des résultats d'une requête). Qu'en penses tu ?

**Comment by FredSafti** [comment]
Tout à fait, ArrayCollection est sufisamment indépendante du reste pour qu'on puisse se poser la question, mais pour moi elle devrait être isolée davantage du reste de la persistence doctrine pour qu'on puisse la voir comme une classe neutre

Oui, j'intègrerais un exemple avec des relations dans la version Doctrine de l'exemple, c'est prévu :)

Pour le reste des questions, je pense surtout qu'il ne faut pas "combattre le framework", mais bien s'en servir. Si je me souviens bien, Doctrine ORM a été créé au départ pour pouvoir justement enregistrer facilement des données correspondant aux notions d'Entity et Repository issues de DDD. Il devrait donc être utilisable pour 99% des cas en ne faisant que du mapping dans le dossier config, sans avoir besoin de passer par des transfert vers entités techniques.

Quand j'évoquais les entités techniques dans mon message plus haut, je parlais de classe qui **extends** celles du domaine, pour justement ne pas avoir à transférer des données d'un objet à un autre. Le code métier ne voit pas qu'il s'agit d'une classe étendue, tandis que la partie Infra manipule ses classes techniques pour load/save les données dans le même objet, donc sans perdre son tracking par l'ORM. On joue à fond sur le LSP de SOLID pour cette partie

Pour aller plus loin, je pense qu'il faut aussi bien différencier les deux besoins : affichage de liste d'entités **différentes** et leurs relations VS affichage / action unitaire sur une entité, avec ou sans ses relations. L'ORM sera complètement adapté au second, il l'est moins pour le premier. Notamment parce qu'on hydrate alors beaucoup d'entités pour ne pas utiliser leur intelligence, c'est donc bien dommage...

Pour ces cas là, je préfère parfois l'utilisation d'agrégats de données. Attention, je ne parle pas des root aggregate de DDD, mais plutôt de DTO qui sont dédiés à la récolte de données dans plusieurs entités, mais destiné à de l'affichage. J'en profite alors pour créer des sortes de Repository pour ces DTO, afin de pouvoir rester plus simple sur ce que j'ai besoin de manipuler dans ces cas là.
Attention par contre, ces DTO doivent absolument être limité à de l'affichage, ils ne doivent porter aucun traitement / intelligence, ni servir de base à un traitement de masse. Cet usage permet également l'agrégation de données plus simplement.
